File: visualization/bean_price.py
import pandas as pd
import matplotlib.pyplot as plt
import argparse
import os

parser = argparse.ArgumentParser(description='Plot the range of bean price for a specified bean.')
parser.add_argument('--name', type=str, default=None, required=True, help='Bean Name, for example Arabica')
args = parser.parse_args()

# Read the sell_price_history.csv data
df = pd.read_csv('database/sell_price_history.csv')

bean_name = args.name.lower()
filtered_data = df[df['source_bean_type'].str.lower() == bean_name]

if filtered_data.shape[0] == 0:
    raise ValueError(f'No bean with name {args.name} found.')

# Convert the month and year columns to a single datetime column
filtered_data['date'] = pd.to_datetime(filtered_data['year'].astype(str) + '-' + filtered_data['month'].astype(str) + '-01')

# Group by the date and calculate min, max, and mean
grouped_data = filtered_data.groupby('date')['price_per_unit'].agg(['min', 'max', 'mean']).reset_index()

# Plot
plt.figure(figsize=(12, 6))

# Fill the region between min and max
plt.fill_between(grouped_data['date'], grouped_data['min'], grouped_data['max'], color='skyblue', alpha=0.4, label='Price Range (Min to Max)')

# Plot mean price
plt.plot(grouped_data['date'], grouped_data['mean'], color='blue', marker='o', label='Mean Price')

plt.title(f'Bean Price Chart for {bean_name}')
plt.xlabel('Time')
plt.ylabel('Price')
plt.legend()
plt.tight_layout()
plt.grid(True, which='both', linestyle='--', linewidth=0.5)

os.makedirs('images', exist_ok=True)
plt.savefig(f"images/bean_price_{bean_name}.png")

File: opti_coffee/visualization/cpp/bean_price.cpp
#include "matplotlibcpp.h"
#include "csv.h"
#include <vector>
#include <string>
#include <map>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <filesystem>
#include <algorithm>
#include <iostream>

namespace plt = matplotlibcpp;

std::map<std::string, std::vector<double>> get_data(const std::string& bean_name) {
    std::map<std::string, std::vector<double>> data;
    io::CSVReader<4> in("database/sell_price_history.csv");
    in.read_header(io::ignore_extra_column, "year", "month", "source_bean_type", "price_per_unit");
    int year, month;
    std::string source_bean_type;
    double price_per_unit;
    while (in.read_row(year, month, source_bean_type, price_per_unit)) {
        if (source_bean_type == bean_name) {
            std::ostringstream date;
            date << year << "-" << std::setfill('0') << std::setw(2) << month << "-01";
            data[date.str()].push_back(price_per_unit);
        }
    }
    return data;
}

void plot_data(const std::map<std::string, std::vector<double>>& data, const std::string& bean_name) {
    std::vector<std::string> x;
    std::vector<double> y_min, y_max, y_mean;
    for (const auto& entry : data) {
        x.push_back(entry.first);
        auto minmax = std::minmax_element(entry.second.begin(), entry.second.end());
        y_min.push_back(*minmax.first);
        y_max.push_back(*minmax.second);
        y_mean.push_back(std::accumulate(entry.second.begin(), entry.second.end(), 0.0) / entry.second.size());
    }

    plt::figure_size(1200, 600);
    plt::fill_between(x, y_min, y_max, "skyblue", 0.4, "Price Range (Min to Max)");
    plt::plot(x, y_mean, "blue", "o", "Mean Price");
    plt::title("Bean Price Chart for " + bean_name);
    plt::xlabel("Time");
    plt::ylabel("Price");
    plt::legend();
    plt::tight_layout();
    plt::grid(true, "both", "both", "--", 0.5);

    std::filesystem::create_directories("images");
    plt::save("images/bean_price_" + bean_name + ".png");
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <bean_name>" << std::endl;
        return 1;
    }

    std::string bean_name = argv[1];
    std::transform(bean_name.begin(), bean_name.end(), bean_name.begin(), ::tolower);

    auto data = get_data(bean_name);
    plot_data(data, bean_name);

    return 0;
}